chain/chain/src/chain.rs

```diff
@@ -3145,6 +3179,34 @@ impl Chain {
         apply_result?;
 
         let shard_state_header = self.get_state_header(shard_id, sync_hash)?;
+        let chunk = shard_state_header.cloned_chunk();
+
+        let block_hash = chunk.prev_block();
+        let block_height = self.get_block_header(block_hash)?.height();
+
+        // Flat storage must not exist at this point because leftover keys corrupt its state.
+        assert!(self.runtime_adapter.get_flat_storage_state_for_shard(shard_id).is_none());
+
+        // We synced shard state on top of _previous_ block for chunk in shard state header and applied state parts to
+        // flat storage. Now we can set flat head to hash of this block and create flat storage.
+        #[cfg(feature = "protocol_feature_flat_state")]
+        {
+            let mut store_update = self.runtime_adapter.store().store_update();
+            store_helper::set_flat_head(&mut store_update, shard_id, block_hash);
+            store_update.commit()?;
+        }
+
+        match self.runtime_adapter.try_create_flat_storage_state_for_shard(
```

Comment:
why do we need to call `try_create_flat_storage_state_for_shard` here? Isn't it true that FlatStorageState is for sure not ready here? Is it just for checking that we are not in the process of flat storage migration? If so, can we add a comment here?

Reply:
To make it less confusing, maybe it is better to split `try_create_flat_storage_state_for_shard` into two methods `check_flat_storage_state_creation_status` and `create_flat_storage_state_for_shard`. Otherwise, it is hard to tell if we are just checking the status here or actually trying to create the flat storage state.

Reply:
What will happen in the following scenario, if we release the code into mainnet, a node starts the flat storage migration process, but the node is also behind for more than an epoch, so it triggers the state sync process as well?

Reply:
I think there is some solution, like "interrupt existing creation process and create everything using state sync". I would even say that DB is corrupted in such case and node owner needs to download a dump. However, DB may not *look* `corrupted for a node owner so they may not even notice a problem...
So it looks like a separate non-trivial question outside of the PR.

Reply:
I think it is ok if state sync and flat storage migration is not compatible, because the migration process only happens once and nodes can sync without state sync (for example, downloading the snapshot). But, let's implement something in the code that explicitly makes the node panic if it tries to state sync when the flat storage migration is in process and tell them to download from the snapshot (which should already have the flat storage ready)